File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/collections/CollectionHandle.java
package in.org.quicko.silo.client.collections;

import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.JsonNode;
import in.org.quicko.silo.client.RequestOptions;
import in.org.quicko.silo.client.entries.Entry;
import in.org.quicko.silo.client.entries.EntryListQuery;
import in.org.quicko.silo.client.entries.EntryMapper;
import in.org.quicko.silo.client.entries.EntryPage;
import in.org.quicko.silo.client.entries.EntryPageStream;
import in.org.quicko.silo.client.entries.EntryReadOptions;
import in.org.quicko.silo.client.entries.EntryReader;
import in.org.quicko.silo.client.entries.EntryStream;
import in.org.quicko.silo.client.pagination.PageWindow;
import in.org.quicko.silo.client.pagination.RowBatch;
import in.org.quicko.silo.client.pagination.RowLoader;
import in.org.quicko.silo.client.scope.RenameOptions;
import in.org.quicko.silo.client.scope.RenameReport;
import in.org.quicko.silo.client.scope.ScopeReference;
import in.org.quicko.silo.client.search.Search;
import in.org.quicko.silo.client.search.SearchPage;
import in.org.quicko.silo.client.search.SearchQuery;
import in.org.quicko.silo.client.search.SearchReach;
import in.org.quicko.silo.client.transport.ApiPath;
import in.org.quicko.silo.client.transport.PagePayload;
import in.org.quicko.silo.client.transport.TransportRequest;
import java.util.List;
import java.util.Map;

/**
 * One collection, typed to its fields: {@code environment.collection("posts",
 * Post.class)}.
 *
 * <p>Reads answer {@link Entry} values and writes take explicit arguments, so
 * there is one entry shape here and no second read surface. Pass
 * {@link EntryReadOptions#raw()} to any read to get the stored
 * {@code {{NAME}}} templates instead of what they resolve to, which is what
 * editing one requires.
 */
public final class CollectionHandle<F> {
  private final ScopeReference scope;
  private final String name;
  private final JavaType fieldsType;
  private final EntryReader reader;
  private final CollectionSchema schema;

  public CollectionHandle(ScopeReference scope, String name, JavaType fieldsType) {
    this.scope = scope;
    this.name = name;
    this.fieldsType = fieldsType;
    this.reader = scope.cache().wrap(new EntryReader(scope, name));
    this.schema = new CollectionSchema(scope, name);
  }

  public String name() {
    return name;
  }

  /** This collection's schema, and the only route that deletes the collection. */
  public CollectionSchema schema() {
    return schema;
  }

  public Entry<F> get(String id) {
    return get(id, EntryReadOptions.none());
  }

  public Entry<F> get(String id, EntryReadOptions options) {
    String path = ApiPath.entry(scope.project(), scope.environment(), name, id);
    options.request().throwIfCancelled("GET", path);
    JsonNode row = scope.transport().codec().tree(reader.get(id, options));
    options.request().throwIfCancelled("GET", path);
    return EntryMapper.read(row, fieldsType, scope.transport().codec());
  }

  public EntryPage<F> list() {
    return list(EntryListQuery.all(), EntryReadOptions.none());
  }

  public EntryPage<F> list(EntryListQuery query) {
    return list(query, EntryReadOptions.none());
  }

  public EntryPage<F> list(EntryListQuery query, EntryReadOptions options) {
    String path = ApiPath.entries(scope.project(), scope.environment(), name);
    options.request().throwIfCancelled("GET", path);
    PagePayload payload = PagePayload.read(scope.transport().codec().tree(reader.list(query, options)));
    options.request().throwIfCancelled("GET", path);
    PageWindow window = new PageWindow(
        payload.limit() == null ? query.limitOrDefault() : payload.limit(),
        payload.offset() == null ? query.offsetOrDefault() : payload.offset());
    List<Entry<F>> entries = payload.rows().stream()
        .map(row -> EntryMapper.<F>read(row, fieldsType, scope.transport().codec())).toList();
    return new EntryPage<>(entries, payload.total(), window,
        next -> list(query.limit(next.limit()).offset(next.offset()), options));
  }

  /** Every matching entry, paged lazily. */
  public EntryStream<F> all() {
    return all(EntryListQuery.all(), EntryReadOptions.none());
  }

  public EntryStream<F> all(EntryListQuery query) {
    return all(query, EntryReadOptions.none());
  }

  public EntryStream<F> all(EntryListQuery query, EntryReadOptions options) {
    RowLoader<Entry<F>> loader = window -> {
      EntryPage<F> page = list(query.limit(window.limit()).offset(window.offset()), options);
      return new RowBatch<>(page.entries(), new PageWindow(page.limit(), page.offset()));
    };
    return new EntryStream<>(loader, query.limitOrDefault(), options.request());
  }

  /** Every matching page, one at a time. */
  public EntryPageStream<F> pages() {
    return pages(EntryListQuery.all(), EntryReadOptions.none());
  }

  public EntryPageStream<F> pages(EntryListQuery query) {
    return pages(query, EntryReadOptions.none());
  }

  public EntryPageStream<F> pages(EntryListQuery query, EntryReadOptions options) {
    return new EntryPageStream<>(() -> list(query, options), options.request());
  }

  public Entry<F> create(F fields) {
    return create(fields, RequestOptions.none());
  }

  public Entry<F> create(F fields, RequestOptions options) {
    return write("POST", ApiPath.entries(scope.project(), scope.environment(), name), fields, null, options);
  }

  /**
   * A full replace, which is what the route is: send every field. {@code rev} is
   * the one the entry answered when it was read, and a stale one raises
   * {@link in.org.quicko.silo.client.errors.ConflictException}.
   */
  public Entry<F> replace(String id, long rev, F fields) {
    return replace(id, rev, fields, RequestOptions.none());
  }

  public Entry<F> replace(String id, long rev, F fields, RequestOptions options) {
    return write("PUT", ApiPath.entry(scope.project(), scope.environment(), name, id), fields, rev, options);
  }

  public void delete(String id, long rev) {
    delete(id, rev, RequestOptions.none());
  }

  public void delete(String id, long rev, RequestOptions options) {
    scope.transport().empty(
        TransportRequest.delete(ApiPath.entry(scope.project(), scope.environment(), name, id))
            .query("rev", rev)
            .options(options)
            .build());
  }

  public SearchPage search(SearchQuery query) {
    return search(query, RequestOptions.none());
  }

  /** Searches this collection only. The reach is the receiver, never an
   *  argument that could be forgotten and widen the search. */
  public SearchPage search(SearchQuery query, RequestOptions options) {
    return new Search(
            scope.transport(),
            SearchReach.collection(scope.project(), scope.environment(), name), reader)
        .run(query, options);
  }

  public RenameReport rename(String newName) {
    return rename(newName, RenameOptions.none());
  }

  public RenameReport rename(String newName, RenameOptions options) {
    return RenameReport.fromWire(scope.transport().json(
        TransportRequest.patch(ApiPath.collection(scope.project(), scope.environment(), name))
            .query("dry_run", options.dryRun() ? true : null)
            .query("expected_id", options.expectedId())
            .body(Map.of("name", newName))
            .options(options.request())
            .build()));
  }

  /** Both writes ask for the stored templates back, so what returns is what was
   *  sent rather than a resolved snapshot of it. */
  private Entry<F> write(String method, String path, F fields, Long rev, RequestOptions options) {
    JsonNode row = scope.transport().json(
        TransportRequest.of(method, path)
            .query("rev", rev)
            .query("variables", "raw")
            .body(fields)
            .options(options)
            .build());
    return EntryMapper.read(row, fieldsType, scope.transport().codec());
  }
}

File: packages/silo-client-dart/lib/src/collections/collection_handle.dart
import '../entries/entry.dart';
import '../entries/entry_list_query.dart';
import '../entries/entry_mapper.dart';
import '../entries/entry_page.dart';
import '../entries/entry_page_stream.dart';
import '../entries/entry_read_options.dart';
import '../entries/entry_reader.dart';
import '../entries/entry_stream.dart';
import '../request_options.dart';
import '../scope/rename_options.dart';
import '../scope/rename_report.dart';
import '../scope/scope_reference.dart';
import '../search/search.dart';
import '../search/search_page.dart';
import '../search/search_query.dart';
import '../search/search_reach.dart';
import '../transport/api_path.dart';
import '../transport/transport_request.dart';
import 'collection_schema.dart';
import 'fields_converter.dart';

/// One collection, its fields read as [F].
///
/// Pass `EntryReadOptions.raw()` to read the stored `{{NAME}}` templates, which
/// editing requires. Every write here drops this collection from the cache.
final class CollectionHandle<F> {
  CollectionHandle(this._scope, this.name, this._converter)
    : schema = CollectionSchema(_scope, name),
      _reader = EntryReader(_scope, name, _converter);

  final ScopeReference _scope;
  final String name;
  final CollectionSchema schema;
  final FieldsConverter<F> _converter;
  final EntryReader<F> _reader;

  /// The same collection with its fields read into, and written from, [R].
  CollectionHandle<R> withConverter<R>({
    required R Function(Map<String, Object?> json) fromJson,
    required Map<String, Object?> Function(R fields) toJson,
  }) => CollectionHandle(_scope, name, FieldsConverter(fromJson: fromJson, toJson: toJson));

  Future<Entry<F>> get(String id, [EntryReadOptions options = const EntryReadOptions()]) => _reader.get(id, options);

  Future<EntryPage<F>> list([
    EntryListQuery query = const EntryListQuery(),
    EntryReadOptions options = const EntryReadOptions(),
  ]) => _reader.list(query, options);

  /// Every matching entry, paged lazily. Not a snapshot of data being written.
  EntryStream<F> all([
    EntryListQuery query = const EntryListQuery(),
    EntryReadOptions options = const EntryReadOptions(),
  ]) => _reader.all(query, options);

  EntryPageStream<F> pages([
    EntryListQuery query = const EntryListQuery(),
    EntryReadOptions options = const EntryReadOptions(),
  ]) => _reader.pages(query, options);

  Future<Entry<F>> create(F fields, [RequestOptions options = const RequestOptions()]) =>
      _write('POST', ApiPath.entries(_scope.project, _scope.environment, name), fields, null, options);

  /// A full replace. [rev] is the one the entry was read with; a stale one is
  /// a `ConflictException`.
  Future<Entry<F>> replace(String id, int rev, F fields, [RequestOptions options = const RequestOptions()]) =>
      _write('PUT', ApiPath.entry(_scope.project, _scope.environment, name, id), fields, rev, options);

  Future<void> delete(String id, int rev, [RequestOptions options = const RequestOptions()]) => _scope.transport.empty(
    TransportRequest(
      'DELETE',
      ApiPath.entry(_scope.project, _scope.environment, name, id),
      query: {'rev': rev},
      options: options,
      evicts: _entries,
    ),
  );

  Future<SearchPage> search(SearchQuery query, [RequestOptions options = const RequestOptions()]) =>
      Search(_scope.transport, SearchReach.collection(_scope.project, _scope.environment, name)).run(query, options);

  Future<RenameReport> rename(String name, [RenameOptions options = const RenameOptions()]) async =>
      RenameReport.fromJson(
        await _scope.transport.json(
          TransportRequest(
            'PATCH',
            ApiPath.collection(_scope.project, _scope.environment, this.name),
            query: {if (options.dryRun) 'dry_run': true, 'expected_id': ?options.expectedId},
            body: {'name': name},
            options: options,
            evicts: _entries,
          ),
        ),
      );

  /// Asks for the stored templates back, so what returns is what was sent.
  Future<Entry<F>> _write(String method, String path, F fields, int? rev, RequestOptions options) async {
    final row = await _scope.transport.json(
      TransportRequest(
        method,
        path,
        query: {'rev': ?rev, 'variables': 'raw'},
        body: _converter.toJson(fields),
        options: options,
        evicts: _entries,
      ),
    );
    return EntryMapper.read(row, _converter);
  }

  String get _entries => ApiPath.entries(_scope.project, _scope.environment, name);
}
